src/spider/client/Driver.hpp

```diff
@@ -0,0 +1,136 @@
+#ifndef SPIDER_CLIENT_DRIVER_HPP
+#define SPIDER_CLIENT_DRIVER_HPP
+
+#include <memory>
+#include <optional>
+#include <string>
+#include <vector>
+
+#include <boost/uuid/uuid.hpp>
+
+#include "../worker/FunctionManager.hpp"
+#include "Job.hpp"
+#include "task.hpp"
+#include "TaskGraph.hpp"
+
+/**
+ * Registers a Task function with Spider
+ * @param func
+ */
+// NOLINTNEXTLINE(cppcoreguidelines-macro-usage)
+#define SPIDER_REGISTER_TASK(func) SPIDER_WORKER_REGISTER_TASK(func)
+
+/**
+ * Registers a timed Task function with Spider
+ * @param func
+ * @param timeout The time after which the task is considered a straggler, triggering Spider to
+ * replicate the task. TODO: Use the timeout.
+ */
+// NOLINTNEXTLINE(cppcoreguidelines-macro-usage)
+#define SPIDER_REGISTER_TASK_TIMEOUT(func, timeout) SPIDER_WORKER_REGISTER_TASK(func)
+
+namespace spider {
+class DriverImpl;
+
+/**
+ * An interface for a client to interact with Spider and create jobs, access the kv-store, etc.
+ */
+class Driver {
+public:
+    /**
+     * @param storage_url
+     */
+    explicit Driver(std::string const& storage_url);
+
+    /**
+     * @param storage_url
+     * @param id A caller-specified ID to associate with this driver. All jobs created by this
+     * driver will be associated with this ID. This may be useful if, for instance, the caller
+     * fails and then needs to reconnect and retrieve all previously created jobs. NOTE: It is
+     * undefined behaviour for two clients to concurrently use the same ID.
+     */
+    Driver(std::string const& storage_url, boost::uuids::uuid id);
```

Comment:
This method should throw an exception if the client ID is already in use, right?